Layertree: only show layers that user is allowed to see

JakobMiksch opened the pull request:
- fixes #562 
- works on login and logout
- also works for groups
- also tested on switch layers
- for testing:
  - set the cookie `roles` to something like `EDITOR_ROLE,DUMMY_ROLE` 
  - add the property `requiredRoles` to layer items in `tree.json`, like below :

```json
[...]
{
            "id": "LA_SITES",
            "text": "Local Authority Sites",
            "leaf": true,
            "requiredRoles": ["READER_ROLE"]
          },
          {
            "id": "RUINS_WFS",
            "leaf": true,
            "text": "Ruins",
            "qtip": "Right-click for a layer grid",
            "requiredRoles": ["ADMIN_ROLE"]
          },
          {
            "id": "my-group",
            "title": "My Group",
            "expanded": true,
            "checked": false,
            "requiredRoles": ["EDITOR_ROLE"],
            "children": [
              {
                "id": "EDGES_WFS",
                "leaf": true,
                "text": "Edges",
                "qtip": "Used in the snapping tool example"
              },
              {
                "id": "NODES_WFS",
                "leaf": true,
                "text": "Nodes",
                "qtip": "Used in the snapping tool example",
                "requiredRoles": ["ADMIN_ROLE"]
              }
            ]
          },
          {
            "id": "BOREHOLE_WMS",
            "leaf": true,
            "text": "Borehole WMS",
            "requiredRoles": ["EDITOR_ROLE"]
          },
          {
            "id": "BOREHOLE_WFS",
            "leaf": true,
            "text": "Borehole WFS (Time)",
            "qtip": "Filter records with the timeslider control",
            "requiredRoles": ["EDITOR_ROLE"]
          },
[...]
```

jansule reviewed (commented):
Very nice @JakobMiksch. Just a few questions:

1. What should happen to a folder, if it does not contain any node a user can see? Should the folder be removed, should it still be visible?
2. Do we have a use case where we have to restore/update the layer tree due to e.g. role changes? If so, we might want to rather filter on the store, instead of actually modifying the original tree.

geographika commented:
Thanks @JakobMiksch !

In relation to the questions from @jansule:

1. Yes it would be good if a folder is empty to hide it from the user. A simpler (and possibly cleaner) solution would to be to allow the `requiredRoles` property to be set on folders directly. 
2. Ideally the function would be rerun if a user logged out and then back in. There is a global event for login and logout that can be listened to: https://github.com/compassinformatics/cpsi-mapview/blob/07e152c4e758f31c554f78125bb92d0eaa5adc58/app/controller/button/LoginButtonController.js#L85

I'm testing the branch on a client application. One thing I noticed is that in `updateTreeChildrenByRole` the following check returns false for all layer nodes as at this stage they are simple JS objects with no `leaf` property (this is only set on the folders in tree.json):

```js
//            if (child.leaf) {
// had to replace with 
            if (child.leaf !== false) {
```

geographika commented:
Just on the `leaf: true` - it looks like we stopped adding this in larger tree.json files, unlike in the demo application.

JakobMiksch commented:
Thanks @geographika @jansule for the review.
- I changed the approach from filtering the JSON to filtering the store. 
- works on login and logout
- also works for groups
- also tested on switch layers
- so far I tested the login/logout with a mock. In `app/controller/form/Login.js`, line 146 I added this:

```js
        Ext.util.Cookies.set('roles', 'EDITOR_ROLE,READER_ROLE');
        Ext.GlobalEvents.fireEvent('login', {});
        view.close();
        return;
```

JakobMiksch commented:
> Just on the `leaf: true` - it looks like we stopped adding this in larger tree.json files, unlike in the demo application.

I did no use the `leaf` in the new approach anymore.

geographika commented:
@JakobMiksch - thanks for this. Looking good when testing in a larger client application. 
One issue is that when a layer is hidden for a user, and then I log out and log in with a user who can see the layer, the layer appears in the tree but is automatically switched on (when it is not configured to be visible in `default.json`. 
I've sent details with an extra login so you may be able to create locally.

geographika commented on app/controller/LayerTreeController.js:
```diff
@@ -202,6 +207,29 @@ Ext.define('CpsiMapview.controller.LayerTreeController', {
+            if (requiredRoles && Ext.isArray(requiredRoles) && requiredRoles.length){
+                var showNode = CpsiMapview.util.RoleManager.hasAtLeastOneRequiredRole(requiredRoles);
+                // needed to ensure layer shown/hidden on both tree and map
+                record.getOlLayer().setVisible(showNode);
```
Not sure if this should be the case - the node should become visible, but only turned on if set in default.json with the:

      "openLayers": {
        "visibility": true
      }

JakobMiksch commented:
> @JakobMiksch - thanks for this. Looking good when testing in a larger client application. One issue is that when a layer is hidden for a user, and then I log out and log in with a user who can see the layer, the layer appears in the tree but is automatically switched on (when it is not configured to be visible in `default.json`. I've sent details with an extra login so you may be able to create locally.

Good Catch! [612e488](https://github.com/compassinformatics/cpsi-mapview/pull/576/commits/612e488c4a2b0018c588ab76a433f6929732eb24) should fix it

geographika commented:
@JakobMiksch - I'm happy to merge this in if you are.

JakobMiksch commented:
@geographika I was waiting for a review by @jansule I will ping him again

jansule commented on app/controller/LayerTreeController.js:
```diff
@@ -202,6 +207,37 @@ Ext.define('CpsiMapview.controller.LayerTreeController', {
 
+
+    /**
+     * Ensures tree and map only contains layers for which
```
```suggestion
     * Ensures tree and map only contain layers for which
```

jansule commented on app/util/RoleManager.js:
```diff
@@ -23,6 +23,22 @@ Ext.define('CpsiMapview.util.RoleManager', {
+     * Checks if user has at least one of the required roles.
+     *
+     * @param {String[]} requiredRoles The required roles.
+     * @return {Boolean} If user has at least one of the required roles.
```
```suggestion
     * @return {Boolean} True, if user has at least one of the required roles. False, otherwise.
```

jansule reviewed (approved):
Very nice @JakobMiksch Just fixed some typos

JakobMiksch commented:
Thanks @jansule for the review, I applied the remarks.
@geographika We can merge from my side.
